swarm/pss/notify/notify.go

```diff
@@ -0,0 +1,244 @@
+package notify
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/swarm/pss"
+)
+
+const (
+	MsgCodeStart = iota
+	MsgCodeNotifyWithKey
+	MsgCodeNotify
+	MsgCodeStop
+	MsgCodeMax
+)
+
+const (
+	DefaultAddressLength = 1
+	minimumAddressLength = 1
+	symKeyLength         = 32 // this should be gotten from source
+)
+
+var (
+	controlTopic = pss.Topic{0x00, 0x00, 0x00, 0x01}
+)
+
+// when code is MsgCodeStart, Payload is address
+// when code is MsgCodeNotify, Payload is symkey. If len = 0, keep old key
+// when code is MsgCodeStop, Payload is address
+type Msg struct {
+	Name    string
+	Payload []byte
+	Code    byte
+}
+
+func (self *Msg) Serialize() []byte {
+	b := bytes.NewBuffer(nil)
+	b.Write([]byte{self.Code})
+	ib := make([]byte, 2)
+	binary.LittleEndian.PutUint16(ib, uint16(len(self.Name)))
+	b.Write(ib)
+	binary.LittleEndian.PutUint16(ib, uint16(len(self.Payload)))
+	b.Write(ib)
+	b.Write([]byte(self.Name))
+	b.Write(self.Payload)
+	return b.Bytes()
+}
+
+func deserializeMsg(msgbytes []byte) (*Msg, error) {
+	msg := &Msg{
+		Code: msgbytes[0],
+	}
+	nameLen := binary.LittleEndian.Uint16(msgbytes[1:3])
+	dataLen := binary.LittleEndian.Uint16(msgbytes[3:5])
+	if int(nameLen+dataLen)+5 != len(msgbytes) {
+		return nil, errors.New("corrupt message")
+	}
+	msg.Name = string(msgbytes[5 : 5+nameLen])
+	msg.Payload = msgbytes[5+nameLen:]
+	return msg, nil
+}
+
+// a notifier has one sendmux entry for each address space it sends messages to
+type sendMux struct {
+	address  pss.PssAddress
+	symKeyId string
+	count    int
+}
+
+type notifier struct {
```

Comment:
Please document the notifier struct. It is not clear what the fields mean. I see that `threshold` is compared to `len` of payload further down, which is not clear when you read the struct.